### DISCUSSION TITLE
Question about scope of `def`

### DISCUSSION THREAD

**Discussion by yohannd1** [janet-lang/janet#1473] [Q&A] [answered]
Hello! I was recently trying to implement a `while-let` macro, and I came across this line in `boot.janet` (implementation of the `if-let` macro):
```janet
          ~(if (def ,bl ,br) ,(aux (+ 2 i)) ,fal2)
```
It made me realize that the scope of variables defined in the condition also extends to its body.

As a more... extreme(?) example:
```janet
(when (> (def x 1) 0)
  (printf "%j is greater than 0" x))
```
Is this intended behavior? I found it really good, but I don't quite understand how this has come to happening.

I have found interesting uses for this, though. I'm experimenting using a sort of "iterator" that eventually returns `nil` when it stops, and I can just throw the def as the condition for a `while`:
```janet
(while (def x (iterator-func))
  (do-something-with x))
```

**Comment by pepe** [comment, marked as answer]
Hello! Yes, from my experience, it is an intended behavior I am using in many places. My guess is that both specials are designed and developed this way intentionally. But I am unsure if I have ever discussed it.

**Reply by sogaiu** [reply]
I don't know if there's a record of this that's easily locatable, but I think we talked about using the return value of `def` in the context of `cond` at least once before (^^;
